### DISCUSSION TITLE
useState/useEffect not working properly without refresh

### DISCUSSION THREAD

**Discussion by sayedulkrm** [vercel/next.js#56903] [Help]
### Summary

I created a Navbar where the logo changes after scrolling. When I deploy to Vercel it doesn't work the first time. After doing multiple refreshes it does work. Am using `App Dir`

**Am not using "use client" on` page.tsx`. Using it only `Navbar.tsx`**
My code:

```
page.tsx

import Footer from "@/components/layout/Footer/Footer";
import Navbar from "@/components/layout/Navbar/Navbar";
import Hero from "@/components/pages/Home/Hero/Hero";
import Section2 from "@/components/pages/Home/Section2/Section2";
import React, { FC } from "react";

type Props = {};

const page: FC<Props> = () => {
    return (
        <div className="w-full h-full  ">
            <Navbar />
            <main>
                <Hero />
                <Section2 />
            </main>
            <Footer />
        </div>
    );
};

export default page;
```

```
Navbar.tsx

"use client";
import React, { useEffect, useState } from "react";
import { usePathname } from "next/navigation";
import Link from "next/link";
import { BiMenuAltRight } from "react-icons/bi";
import Image from "next/image";

import whitelogo from "../../../../public/assets/whitelogo.png";

import colorLogo from "../../../../public/assets/logo.png";

type Props = {};

const Navbar = () => {
    const pathname = usePathname();

    const [isSticky, setIsSticky] = useState(false);
    const [mounte, setMounte] = useState(false);

    const [open, setOpen] = useState(false);
    const showDrawer = () => {
        setOpen(true);
    };
    const onClose = () => {
        setOpen(false);
    };

    useEffect(() => {
        const handleScroll = () => {
            const offset = window.scrollY;
            if (offset > 40) {
                setIsSticky(true);
            } else {
                setIsSticky(false);
            }
        };

        window.addEventListener("scroll", handleScroll);

        return () => {
            window.removeEventListener("scroll", handleScroll);
        };
    }, []);

return(

<div>
                    <Link href={"/"} className="w-auto">
                        <Image
                            src={ isSticky ? whitelogo  :  colorLogo}
                            // width={100}
                            alt="logistic"
                            className="h-auto w-20 md:w-32"
                        />
                    </Link>
                </div>


)
}

```

Thank you

**Comment by icyJoseph** [comment]
Hi,

Got a link to the deployed code? Or repository?

Do you maybe have a service worker or something?

By the way, you can synchronize to the scroll event, and a particular value by using `useSyncExternalStore`:

- https://thisweekinreact.com/articles/useSyncExternalStore-the-underrated-react-api#another-example-scrolly

**Reply by sayedulkrm** [reply]
Those other components just have a default 
`<div>Element Name</div>`
Nothing extra. I actually just want to make it SSR. that's why don't want to add "use client" in the page.tsx

**Reply by icyJoseph** [reply]
Yeah but that shouldn't matter. NavBar having the client directive is enough.

Do you have a repository/link to project to analyze? With this much I'm not sure there's anything to do. I'll see if I spot something obvious on the examples you provide.

**Reply by sayedulkrm** [reply]
https://github.com/sayedulkrm/anas-project-1
This one

**Reply by sayedulkrm** [reply]
I did some updated. Please check now

**Reply by icyJoseph** [reply]
Alright, what if you do:

```ts
  useEffect(() => {
    const handleScroll = () => {
      const offset = window.scrollY;
      if (offset > 40) {
        setIsSticky(true);
      } else {
        setIsSticky(false);
      }
    };

    window.addEventListener('scroll', handleScroll);

    handleScroll(); // ---> this part

    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, []);
```

So that when you sync with the current scroll immediately.

**Reply by icyJoseph** [reply]
Also as shown in:

- https://thisweekinreact.com/articles/useSyncExternalStore-the-underrated-react-api#another-example-scrolly

You could just do this:

```tsx
'use client';
import React, { useEffect, useState, useSyncExternalStore } from 'react';
import { usePathname } from 'next/navigation';
import Link from 'next/link';
import { BiMenuAltRight } from 'react-icons/bi';
import Image from 'next/image';

import whitelogo from '../../../../public/assets/whitelogo.png';

import colorLogo from '../../../../public/assets/logo.png';

function subscribe(onStoreChange: VoidFunction) {
  global.window?.addEventListener('scroll', onStoreChange);
  return () => global.window?.removeEventListener('scroll', onStoreChange);
}

function useScrollY(): number;
function useScrollY<Result>(selector?: (offset: number) => Result): Result;
function useScrollY(selector = (id: number) => id) {
  return useSyncExternalStore(
    subscribe,
    () => selector(global.window?.scrollY),
    () => selector(0)
  );
}

const Navbar = () => {
  // other stuff goes here

  const isSticky = useScrollY((offset) => offset > 40);

  return (
    <header
      className={`${
        isSticky && pathname === '/' ? 'stick' : 'navbar !text-white'
      } ${pathname !== '/' && 'stick'} py-3 rounded-md px-6 font-roboto`}
    >
      {/* rest of UI */}
    </header>
  );
};

export default Navbar;

```

And than can lead to improved rendering.

**Reply by sayedulkrm** [reply]
I use both of your code and it's still the same. It takes some time to change the logo when scrolling. It doesn't change immediately. Can you please tell me why it doesn't change immediately?
You can open the link in Incognito Mode

**Reply by icyJoseph** [reply]
Could you record a video or something of what do you mean? You mean it starts "not-sticky" even though it is clearly further down? 

Try using this otherwise:

```ts
import { useEffect, useLayoutEffect } from 'react'

export const useIsomorphicLayoutEffect =
  typeof window !== 'undefined' ? useLayoutEffect : useEffect
```

And use `useIsomorphicLayoutEffect` instead of `useEffect`.

**Reply by sayedulkrm** [reply]
Yeah, that also. I'll make a video and show you.

1) initially when we open the link in a new browser if you scroll down immediately then the "color logo" now shows. It still shows the "white logo"
2) sometimes the navbar is not sticky (stick to top) too.

those issues are solved by themselves after some time. If you can https://anas-project-1.vercel.app/ open this link in the new browser (where the link previously did not open) then immediately scroll down.

**Reply by sayedulkrm** [reply]
This is a issue though? Or is it not?

**Reply by icyJoseph** [reply]
It doesn't look like an issue just yet. I'm not sure what it is you are seeing vs expecting.

Please try to record a video.

**Reply by sayedulkrm** [reply]
Please check the video
https://www.loom.com/share/c96bdf53e06142fc84bdf07c36c5f584?sid=2547fbd1-b6cb-4ea6-a91a-9e7d393960cf

As you can see in `"Navbar"` when I scroll the logo not changing. The logo stays `"WHITE"`. Though it should be `"Color"` Logo

**Reply by sayedulkrm** [reply]
Is that my code issue? or is it next Js loading issue?

**Reply by icyJoseph** [reply]
Oh wow, I don't see that at all.

I'm a bit busy atm, but let's try:

- incognito window
- make sure the CSS to change colors exists. To do this, inspect the DOM and stylesheets on the browser

Post back screenshots if possible

**Reply by sayedulkrm** [reply]
Okay please let me know once you check that. It's not color-changing. It's directly an image. There are 2 images. 1 is a `Colored image `and another is a `White color`. So basically there is if `isSticky ? colorImage : whitecolor image`
Also am using `next/image` tag. `<Image />`. Should I use `<img />` tag?

**Reply by icyJoseph** [reply]
Try like this:

```ts
src={isSticky ? "/assets/whitelogo.png" : "/assets/logo.png"}
```

Make sure to get the second one correctly. I don't remember what's called in your app :)

**Reply by sayedulkrm** [reply]
Here i define that on top

```
import whitelogo from "../../../../public/assets/whitelogo.png";

import colorLogo from "../../../../public/assets/logo.png";
...

Then the image is

    <div>
                    <Link href={"/"} className="w-auto">
                        <Image
                            src={isSticky ? colorLogo : whitelogo}
                            // width={100}
                            alt="logistic"
                            className="h-auto w-20 md:w-32"
                        />
                    </Link>
                </div>



```

**Reply by icyJoseph** [reply]
Yeah but you are using something called a static path object as a source, but you can also actually just refer to the actual `/assets/whitelogo.png`. It is not the same thing.

**Reply by sayedulkrm** [reply]
I did. But it's still the same issue. Please open the link in the `incognito window`
https://anas-project-1.vercel.app/

It gets fixed by it self after 10-15s. 
Is it my code issue? or the nextjs issue?

**Reply by icyJoseph** [reply]
I am starting to think that it might be a network speed issue. The image takes a while to download, and that's why "waiting" fixes it by itself. 

Could you inspect the network tab, and see how long it takes to download the images?

**Reply by sayedulkrm** [reply]
Yeah, I just noticed now. It keeps getting pending whenever I scroll. Is there any way to fix it?
![image](https://github.com/vercel/next.js/assets/100830568/0c31689a-e3ed-4b29-b3f3-dce37b3fe361)

**Reply by icyJoseph** [reply]
Mmm I guess that as you scroll, you change the request of the image.

Could it be that you are located far way from where your app is hosted? 

- https://vercel.com/guides/choosing-deployment-regions

Maybe you can change the settings in the Vercel Dashboard for your project? So that it is physically closer to you? 6.76 seconds to download an image is way too much.

**Reply by sayedulkrm** [reply]
But basically, it doesn't give any problem with a normal react website. Is there any way I can stop it from being downloaded every time?

right now my asset folder is in `public dir`. If I move it in `src` can it solve?

**Reply by icyJoseph** [reply]
I mean, try to give the image priority, because the next image component does lazy loading. 

It has nothing to do with React.

It is like, the next image also runs optimizations server side. Maybe also try to pass unoptimized set to true.

**Reply by sayedulkrm** [reply]
Add the priority and nothing changes :(

**Reply by sayedulkrm** [reply]
Is that my code issue or network issue or nextjs issue?
